src/compiler/checker.ts

```diff
@@ -7771,6 +7771,11 @@ namespace ts {
                 if (target.flags & TypeFlags.Union && containsType(targetTypes, source)) {
                     return Ternary.True;
                 }
+                const discriminantType = findMatchingDiscriminantType(source, target);
+                if (discriminantType) {
+                    return isRelatedTo(source, discriminantType, reportErrors);

```

Comment:
How would the recursive call of `isRelatedTo` know that it was comparing a discriminated union member to a discriminant? I guess `reportErrors` could take on a third value that means "only report errors if you can find a matching discriminant", but that seems like using a general-purpose flag for a special case.